File: frontend/apps/web-antd/src/api/tuning.ts
/**
 * 整定模块 API（09 设计方案恢复为一级模块，2026-08-19）
 *
 * 设计文档：docs/MVP设计/09-整定模块设计方案.md §5.2/§5.3
 * 后端：backend/app/api/v1/endpoints/tuning.py（/api/v1/tuning/*）
 */

import { requestClient } from '#/api/request';

export namespace TuningApi {
  /** 模型类型 */
  export type ModelType = 'FOPDT' | 'IPDT' | 'SOPDT';
  /** 整定算法（矩阵 5 算法 + 手动整定） */
  export type TuningAlgorithm =
    | 'COHEN_COON'
    | 'IDENTIFICATION_ONLY'
    | 'IMC'
    | 'LAMBDA'
    | 'MANUAL_TUNING'
    | 'SIMC'
    | 'ZN';
  /** 可信度等级 */
  export type ConfidenceLevel = 'A' | 'B' | 'C' | 'D' | 'E' | 'INCONCLUSIVE';
  /** 整定任务状态（MVP 实际使用 SIMULATED→APPLIED→VERIFIED 路径） */
  export type TuningTaskStatus =
    | 'APPLIED'
    | 'COMPLETED'
    | 'DRAFT'
    | 'IDENTIFIED'
    | 'INCONCLUSIVE'
    | 'PENDING'
    | 'ROLLED_BACK'
    | 'RUNNING'
    | 'SIMULATED'
    | 'VERIFIED';

  /** 模型参数（FOPDT: K/tau/theta；SOPDT 标准形: K/T1/T2/theta） */
  export interface ModelParams {
    K?: null | number;
    T1?: null | number;
    T2?: null | number;
    tau?: null | number;
    theta?: null | number;
  }

  export interface PidParams {
    kp: number;
    ti: number;
    td: number;
  }

  export interface PidParamsWithLabel extends PidParams {
    label: string;
  }

  /** 阶跃实验辨识结果 */
  export interface ModelIdentifyResult {
    modelType: ModelType;
    params: ModelParams;
    fittingScore: number;
    algorithmVersion: string;
    dataPoints: number;
    recordId?: null | string;
    fittedCurve?: null | { [key: string]: any[] };
  }

  /** 历史数据辨识异步响应 */
  export interface IdentifyHistoryAsyncResponse {
    taskId: string;
    [key: string]: any;
  }

  /** 异步任务进度 */
  export interface TaskProgress {
    taskId: string;
    status: 'FAILED' | 'PENDING' | 'RUNNING' | 'SUCCESS';
    progress: number;
    stage?: null | string;
    message?: null | string;
    result?: null | {
      [key: string]: any;
      confidenceLevel?: ConfidenceLevel;
      fittingScore?: number;
      modelType?: ModelType;
      params?: ModelParams;
      reason?: string;
    };
    error?: null | string;
  }

  /** 整定方法信息 */
  export interface TuningMethodInfo {
    code: string;
    name: string;
    description: string;
    applicableModel: string;
    params: Record<string, any>[];
  }

  /** 单算法整定结果 */
  export interface TuneResult {
    algorithm: string;
    recommendedPid: PidParams;
    currentPid?: null | PidParams;
    algorithmParams?: null | Record<string, any>;
    algorithmVersion: string;
    notes?: null | string;
    risk?: null | {
      description?: null | string;
      factors: string[];
      riskLevel: string;
    };
    rollbackPid?: null | PidParams;
  }

  /** 全算法矩阵行（09 §4.2：单行失败不阻断） */
  export interface TuneMatrixRow {
    algorithm: TuningAlgorithm;
    ok: boolean;
    result?: TuneResult;
    error?: string;
  }

  /** 仿真性能指标 */
  export interface SimulationMetrics {
    riseTime?: null | number;
    overshoot?: null | number;
    settlingTime?: null | number;
    itae?: null | number;
  }

  /** 候选 PID 响应（多 PID 对比） */
  export interface CandidateResponse {
    label: string;
    response: { op: number[]; pv: number[]; sp: number[] };
    metrics: SimulationMetrics;
  }

  /** 闭环仿真结果 */
  export interface SimulationResult {
    timestamps: number[];
    currentResponse: { op: number[]; pv: number[]; sp: number[] };
    recommendedResponse: { op: number[]; pv: number[]; sp: number[] };
    currentMetrics: SimulationMetrics;
    recommendedMetrics: SimulationMetrics;
    improvement: Record<string, null | number>;
    candidateResponses?: CandidateResponse[] | null;
  }

  /** 整定任务列表项 */
  export interface TuningTaskItem {
    id: string;
    loopId: string;
    tagName?: null | string;
    modelType: ModelType;
    modelParams?: null | Record<string, any>;
    algorithm: string;
    recommendedPid?: null | PidParams;
    fittingScore?: null | number;
    status: TuningTaskStatus;
    createdBy?: null | string;
    createdAt: string;
    confidenceLevel?: ConfidenceLevel | null;
    confidenceReason?: null | string;
    identifyMethod?: null | string;
    dataSource?: null | string;
    excitationScore?: null | number;
    residualTestPassed?: boolean | null;
    taskId?: null | string;
    completedAt?: null | string;
  }

  export interface TuningTaskDetail extends TuningTaskItem {
    simulationResult?: null | Record<string, any>;
    currentPid?: null | PidParams;
  }

  export interface TuningTaskListData {
    items: TuningTaskItem[];
    total: number;
    page: number;
    pageSize: number;
  }

  export interface TuningHistoryStats {
    totalTasks: number;
    byAlgorithm: Record<string, number>;
    byStatus: Record<string, number>;
    avgFittingScore?: null | number;
    recentTasks: TuningTaskItem[];
  }

  /** KPI 快照摘要（与处置模块同口径） */
  export interface KpiSummary {
    score: null | number;
    goodValueRate: null | number;
    effectiveAutoRate: null | number;
    steadyRate: null | number;
    accuracyRate: null | number;
    fastRate: null | number;
    oscillationRate: null | number;
    saturationRate: null | number;
    confidenceLevel: null | string;
    tsStart: null | string;
    tsEnd: null | string;
  }

  /** 波形序列（get_waveform 契约） */
  export interface WaveformData {
    loopId: string;
    tagName?: string;
    timestamps: string[];
    pv: (null | number)[];
    sp: (null | number)[];
    op: (null | number)[];
    mode?: (null | number)[];
    pvQuality?: string[];
    downsampled?: boolean;
    pointCount?: number;
    sampleInterval?: number;
  }

  /** 效果验证前后窗数据（09 §4.5） */
  export interface VerificationData {
    loopId: string;
    pointTime: string;
    windowHours: number;
    before: WaveformData;
    after: WaveformData;
    kpiBefore: KpiSummary | null;
    kpiAfter: KpiSummary | null;
    /** 后窗超出当前时刻（数据截至当前时刻） */
    afterTruncated: boolean;
  }

  /** 整定批次状态（status 为 B-06 动态阻塞判定后的有效状态） */
  export type TuningBatchStatus =
    | 'BLOCKED'
    | 'CANCELLED'
    | 'COMPLETED'
    | 'PENDING'
    | 'READY'
    | 'RUNNING';

  /** 整定批次列表摘要（追溯矩阵 GAP-2a） */
  export interface TuningBatchSummary {
    id: number;
    batchNo: string;
    title: string;
    scopeType: string;
    scopeId: number;
    status: TuningBatchStatus;
    storedStatus: string;
    blocked: boolean;
    blockReason?: null | string;
    recordCount: number;
    createdAt?: null | string;
  }

  export interface TuningBatchListData {
    items: TuningBatchSummary[];
    total: number;
    page: number;
    pageSize: number;
  }

  /** 批次前置工单摘要 */
  export interface TuningBatchPrereqOrder {
    orderId: string;
    orderNo?: null | string;
    title?: null | string;
    status?: null | string;
    closed: boolean;
  }

  /** 批次关联整定记录（tuning_batch_records N:M） */
  export interface TuningBatchRecordItem {
    recordId: string;
    sortOrder: number;
    loopId: string;
    tagName?: null | string;
    modelType?: null | string;
    algorithm?: null | string;
    status?: null | string;
    fittingScore?: null | number;
    createdBy?: null | string;
    createdAt?: null | string;
  }

  /** 整定批次详情 */
  export interface TuningBatchDetail extends TuningBatchSummary {
    prereqOrderIds: string[];
    prereqOrders: TuningBatchPrereqOrder[];
    /** scatters JSONB 原样返回（[{loop_id, score, ...}]） */
    scattersBefore?: null | Record<string, any>[];
    scattersAfter?: null | Record<string, any>[];
    ownerId?: null | number;
    expectedStartAt?: null | string;
    actualStartAt?: null | string;
    completedAt?: null | string;
    records: TuningBatchRecordItem[];
  }
}

/** 整定方法信息 */
export function getTuningMethodsApi() {
  return requestClient.get<TuningApi.TuningMethodInfo[]>('/tuning/methods');
}

/** 阶跃实验辨识（同步） */
export function identifyStepApi(data: {
  endTime: string;
  loopId: string;
  method?: string;
  modelType?: TuningApi.ModelType;
  startTime: string;
}) {
  return requestClient.post<TuningApi.ModelIdentifyResult>(
    '/tuning/identify',
    data,
  );
}

/** 历史数据辨识（异步，返回 taskId 轮询） */
export function identifyHistoryApi(data: {
  candidateModelTypes?: TuningApi.ModelType[];
  endTime: string;
  loopId: string;
  startTime: string;
  thetaEstimate?: null | number;
}) {
  return requestClient.post<TuningApi.IdentifyHistoryAsyncResponse>(
    '/tuning/identify/history',
    data,
  );
}

/** 异步任务进度查询（taskId 为 Celery 任务 ID） */
export function getTuningTaskStatusApi(taskId: string) {
  return requestClient.get<TuningApi.TaskProgress>(
    `/tuning/tasks/${taskId}/status`,
  );
}

/** 取消异步任务 */
export function cancelTuningTaskApi(taskId: string) {
  return requestClient.post<{ cancelled: boolean }>(
    `/tuning/tasks/${taskId}/cancel`,
  );
}

/** 单算法整定（矩阵行内参数微调重算） */
export function tuneSingleApi(data: {
  algorithm: TuningApi.TuningAlgorithm;
  algorithmParams?: null | Record<string, any>;
  currentPid?: null | TuningApi.PidParams;
  loopId?: null | string;
  modelParams: TuningApi.ModelParams;
  modelSource?: string;
  modelType: TuningApi.ModelType;
  riskConfirmed?: boolean;
  sourceRecordId?: null | string;
}) {
  return requestClient.post<TuningApi.TuneResult>('/tuning/tune', data);
}

/** 全算法矩阵（5 算法一次全算，09 §4.2） */
export function tuneMatrixApi(data: {
  algorithmParams?: null | Record<string, any>;
  currentPid?: null | TuningApi.PidParams;
  loopId?: null | string;
  modelParams: TuningApi.ModelParams;
  modelSource?: string;
  modelType: TuningApi.ModelType;
  riskConfirmed?: boolean;
  sourceRecordId?: null | string;
}) {
  return requestClient.post<{ rows: TuningApi.TuneMatrixRow[] }>(
    '/tuning/tune/matrix',
    data,
  );
}

/** 多 PID 对比仿真（pidCandidates ≥2 组） */
export function comparePidsApi(data: {
  currentPid?: null | TuningApi.PidParams;
  loopId?: null | string;
  modelParams: TuningApi.ModelParams;
  modelSource?: string;
  modelType: TuningApi.ModelType;
  pidCandidates: TuningApi.PidParamsWithLabel[];
  riskConfirmed?: boolean;
  setpointStep?: number;
  simDuration?: number;
  simStep?: number;
  sourceRecordId?: null | string;
}) {
  return requestClient.post<TuningApi.SimulationResult>(
    '/tuning/compare',
    data,
  );
}

/** 保存整定方案（09 §4.4：显式保存才落记录） */
export function saveTuningTaskApi(data: {
  [key: string]: any;
  algorithm: string;
  currentPid?: null | TuningApi.PidParams;
  fittingScore?: null | number;
  loopId: string;
  modelParams: TuningApi.ModelParams;
  modelType: TuningApi.ModelType;
  recommendedPid: TuningApi.PidParams;
  simulationResult?: null | Record<string, any>;
  status?: TuningApi.TuningTaskStatus;
}) {
  return requestClient.post<{ id: string }>('/tuning/tasks', data);
}

/** 整定任务列表（分页 + 筛选；status 支持逗号多值，startTime/endTime 按创建时间过滤） */
export function getTuningTasksApi(params: {
  algorithm?: string;
  endTime?: string;
  loopId?: string;
  page?: number;
  pageSize?: number;
  startTime?: string;
  status?: string;
}) {
  return requestClient.get<TuningApi.TuningTaskListData>('/tuning/tasks', {
    params,
  });
}

/** 整定任务详情 */
export function getTuningTaskDetailApi(taskId: string) {
  return requestClient.get<TuningApi.TuningTaskDetail>(
    `/tuning/tasks/${taskId}`,
  );
}

/** 整定历史统计 */
export function getTuningHistoryApi() {
  return requestClient.get<TuningApi.TuningHistoryStats>('/tuning/history');
}

/** 整定批次列表（分页 + status/创建时间窗筛选，追溯矩阵 GAP-2a） */
export function getTuningBatchesApi(params: {
  endTime?: string;
  page?: number;
  pageSize?: number;
  startTime?: string;
  status?: string;
}) {
  return requestClient.get<TuningApi.TuningBatchListData>('/tuning/batches', {
    params,
  });
}

/** 整定批次详情（关联整定记录 + 前置工单摘要 + scatters） */
export function getTuningBatchDetailApi(batchId: number | string) {
  return requestClient.get<TuningApi.TuningBatchDetail>(
    `/tuning/batches/${batchId}`,
  );
}

/** 效果验证前后窗曲线数据（09 §4.5，实时拉取不落库） */
export function getVerificationDataApi(params: {
  loopId: string;
  pointTime: string;
  windowHours: number;
}) {
  return requestClient.get<TuningApi.VerificationData>(
    '/tuning/verification/data',
    { params },
  );
}

File: backend/app/schemas/tuning.py
"""Tuning center schemas (IDS v3.2 §2.5 — S7-TUNE-006).

对齐关键算法设计说明 v1.0 §6：
- FOPDT/SOPDT/IPDT 模型辨识
- IMC/Lambda/Z-N/Cohen-Coon/SIMC PID 整定
- 闭环仿真（RK4 + 增量式 PID）

Phase 2 扩展（2026-07-28）：
- 历史数据辨识（identifyStrategy/candidateModelTypes/confidenceLevel）
- 多 PID 对比（pidCandidates/candidateResponses）
- 异步任务（TaskProgress）
- 状态机对齐实现契约（DRAFT/RUNNING/IDENTIFIED/SIMULATED/COMPLETED/INCONCLUSIVE/ROLLED_BACK）
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from app.schemas.base import CamelModel

# ---------------------------------------------------------------------------
# 枚举类型定义（S4-C3）
# 与数据库 CheckConstraint 保持一致（app/models/tuning.py）
# ---------------------------------------------------------------------------

# 模型类型：FOPDT/SOPDT/IPDT
ModelType = Literal["FOPDT", "SOPDT", "IPDT"]

# 历史算法当前仅对 FOPDT/SOPDT 有真实转换链；IPDT 保留在阶跃实验路径。
HistoryModelType = Literal["FOPDT", "SOPDT"]

# 整定算法：IMC/LAMBDA/ZN/COHEN_COON/SIMC
TuningAlgorithm = Literal["IMC", "LAMBDA", "ZN", "COHEN_COON", "SIMC"]

# 整定任务状态（Phase 2 新枚举 + 兼容旧枚举）
TuningTaskStatus = Literal[
    # Phase 2 新枚举
    "DRAFT",
    "RUNNING",
    "IDENTIFIED",
    "SIMULATED",
    "COMPLETED",
    "INCONCLUSIVE",
    "ROLLED_BACK",
    # 旧枚举（兼容期保留）
    "PENDING",
    "APPLIED",
    "VERIFIED",
]

# 辨识策略
IdentifyStrategy = Literal["AUTO", "HISTORY_ONLY", "STEP_ONLY"]

# 辨识方法
IdentifyMethod = Literal[
    "HISTORICAL_ARX",
    "HISTORICAL_ARMAX",
    "HISTORICAL_IV",
    "STEP_TWO_POINT",
    "STEP_AREA",
    "STEP_NLS",
]

# 数据来源（fallback_step = AUTO 策略历史辨识失败/数据不足后的阶跃兜底标记，P1-6）
DataSource = Literal["HISTORY", "STEP_EXPERIMENT", "fallback_step"]

# 可信度等级
ConfidenceLevel = Literal["A", "B", "C", "D", "E", "INCONCLUSIVE"]

# 纯滞后参数来源
ThetaSource = Literal["EXPLICIT", "HEURISTIC_2TS"]

# 推荐链模型来源。STEP_EXPERIMENT 只能由服务端验证过的阶跃记录/内部链路放行。
ModelSource = Literal["IDENTIFICATION_RECORD", "STEP_EXPERIMENT", "MANUAL"]


# ---------------------------------------------------------------------------
# 模型辨识
# ---------------------------------------------------------------------------


class ModelIdentifyRequest(CamelModel):
    """POST /tuning/identify 请求体（阶跃实验路径，保留向后兼容）。"""

    loopId: str = Field(..., description="回路 ID")
    startTime: str = Field(..., description="起始时间 ISO 8601")
    endTime: str = Field(..., description="结束时间 ISO 8601")
    modelType: ModelType = Field("FOPDT", description="模型类型: FOPDT/SOPDT/IPDT")
    method: str | None = Field(
        None, description="辨识方法: TWO_POINT/AREA（仅 FOPDT，默认 TWO_POINT）"
    )


class ModelIdentifyHistoryRequest(CamelModel):
    """POST /tuning/identify/history 请求体（Phase 2 历史数据辨识路径）."""

    loopId: str = Field(..., description="回路 ID")
    startTime: str = Field(..., description="起始时间 ISO 8601")
    endTime: str = Field(..., description="结束时间 ISO 8601")
    identifyStrategy: IdentifyStrategy = Field(
        "AUTO", description="辨识策略: AUTO(优先历史,失败兜底阶跃)/HISTORY_ONLY/STEP_ONLY"
    )
    candidateModelTypes: list[HistoryModelType] | None = Field(
        None, description="候选模型阶次列表，默认 [FOPDT, SOPDT]"
    )
    thetaEstimate: float | None = Field(
        None,
        ge=0,
        description="纯滞后预估值（秒）；None 使用 2Ts 启发值并将可信度封顶 C",
    )


class ModelParams(CamelModel):
    """模型参数。"""

    K: float | None = Field(None, description="过程增益")
    tau: float | None = Field(None, description="时间常数（秒）")
    theta: float | None = Field(None, description="死区时间（秒）")
    T1: float | None = Field(None, description="SOPDT 第一时间常数（秒）")
    T2: float | None = Field(None, description="SOPDT 第二时间常数（秒）")


class CandidateModel(CamelModel):
    """候选模型（多阶次并行辨识结果之一）。"""

    modelType: ModelType
    params: ModelParams
    fittingScore: float = Field(..., description="拟合度 R²（%）")
    confidence: ConfidenceLevel
    identifyMethod: IdentifyMethod | None = None
    residualTestPassed: bool | None = None
    excitationScore: float | None = None
    reason: str | None = None


class ModelIdentifyResult(CamelModel):
    """模型辨识结果（阶跃实验路径）。"""

    modelType: ModelType
    params: ModelParams
    fittingScore: float = Field(..., description="拟合度 R²（%）")
    algorithmVersion: str
    dataPoints: int = Field(..., description="参与辨识的数据点数")
    recordId: str | None = Field(None, description="服务端持久化的阶跃辨识记录 ID")
    # 拟合曲线（用于前端可视化）
    fittedCurve: dict[str, list[Any]] | None = Field(
        None, description="拟合曲线 {timestamps: [], pv: [], fitted: []}"
    )


class ModelIdentifyHistoryResult(CamelModel):
    """历史数据辨识结果（Phase 2 路径）."""

    success: bool
    modelType: str | None = None
    params: dict[str, Any] | None = None
    fittingScore: float | None = Field(None, description="拟合度 R²（%）")
    confidenceLevel: ConfidenceLevel | None = None
    dataConfidenceLevel: ConfidenceLevel | None = Field(
        None, description="数据质量可信度（基于 valid_rate）"
    )
    confidenceReason: str | None = None
    thetaSource: ThetaSource | None = Field(None, description="纯滞后参数来源")
    excitationScore: float | None = None
    residualTestPassed: bool | None = None
    identifyMethod: IdentifyMethod | None = None
    candidateModels: list[CandidateModel] | None = None
    algorithmVersion: str | None = None
    dataPoints: int | None = None
    validRate: float | None = Field(None, description="有效数据率 0~1")
    samplingFreq: float | None = Field(None, description="采样频率（Hz）")
    reason: str | None = None
    tagName: str | None = None


# ---------------------------------------------------------------------------
# PID 整定
# ---------------------------------------------------------------------------


class PidParams(CamelModel):
    """PID 参数。"""

    kp: float = Field(..., description="比例增益")
    ti: float = Field(..., description="积分时间（秒）")
    td: float = Field(0.0, description="微分时间（秒）")


class PidParamsWithLabel(CamelModel):
    """带标签的 PID 参数（用于多 PID 对比）."""

    label: str = Field(..., description="PID 标签（如 IMC λ=1.0）")
    kp: float = Field(..., description="比例增益")
    ti: float = Field(..., description="积分时间（秒）")
    td: float = Field(0.0, description="微分时间（秒）")


class TuneRequest(CamelModel):
    """POST /tuning/tune 请求体。"""

    modelType: ModelType = Field(..., description="模型类型: FOPDT/SOPDT/IPDT")
    modelParams: ModelParams
    algorithm: TuningAlgorithm = Field(..., description="整定算法: IMC/LAMBDA/ZN/COHEN_COON/SIMC")
    algorithmParams: dict[str, Any] | None = Field(
        None, description="算法参数（如 lambda 比例系数）"
    )
    currentPid: PidParams | None = Field(None, description="当前 PID 参数（用于对比）")
    loopId: str | None = Field(None, description="回路 ID（可选，用于记录）")
    sourceRecordId: str | None = Field(None, description="模型辨识记录 ID")
    modelSource: ModelSource | None = Field(
        None,
        description="模型来源；旧请求可解析但不会绕过服务端安全门禁",
    )
    riskConfirmed: bool = Field(False, description="是否已显式确认 C 级/人工模型风险")


class TuneResult(CamelModel):
    """PID 整定结果。"""

    algorithm: TuningAlgorithm
    recommendedPid: PidParams
    currentPid: PidParams | None = None
    algorithmParams: dict[str, Any] | None = None
    algorithmVersion: str
    notes: str | None = Field(None, description="整定说明")


# ---------------------------------------------------------------------------
# 闭环仿真
# ---------------------------------------------------------------------------


class SimulateRequest(CamelModel):
    """POST /tuning/simulate 请求体。"""

    modelType: ModelType = Field("FOPDT", description="模型类型")
    modelParams: ModelParams
    currentPid: PidParams
    recommendedPid: PidParams
    pidCandidates: list[PidParamsWithLabel] | None = Field(
        None, description="多组候选 PID 参数（Phase 2 新增，向后兼容）"
    )
    simDuration: float = Field(600.0, description="仿真时长（秒）")
    simStep: float = Field(1.0, description="仿真步长（秒）")
    setpointStep: float = Field(1.0, description="设定值阶跃幅值")
    disturbanceType: str = Field("step", description="扰动类型: step/none")
    loopId: str | None = Field(None, description="回路 ID（模型记录校验用）")
    sourceRecordId: str | None = Field(None, description="模型辨识记录 ID")
    modelSource: ModelSource | None = Field(
        None,
        description="模型来源；旧请求可解析但不会绕过服务端安全门禁",
    )
    riskConfirmed: bool = Field(False, description="是否已显式确认 C 级/人工模型风险")


class SimulationMetrics(CamelModel):
    """仿真性能指标。"""

    riseTime: float | None = Field(None, description="上升时间（秒）")
    overshoot: float | None = Field(None, description="超调量（%）")
    settlingTime: float | None = Field(None, description="稳定时间（秒）")
    itae: float | None = Field(None, description="ITAE 积分")


class CandidateResponse(CamelModel):
    """候选 PID 响应（多 PID 对比）."""

    label: str = Field(..., description="PID 标签")
    response: dict[str, list[float]]
    metrics: SimulationMetrics


class SimulationResult(CamelModel):
    """闭环仿真结果。"""

    timestamps: list[float]
    currentResponse: dict[str, list[float]]
    recommendedResponse: dict[str, list[float]]
    currentMetrics: SimulationMetrics
    recommendedMetrics: SimulationMetrics
    improvement: dict[str, float | None]
    # Phase 2 新增：多 PID 对比
    candidateResponses: list[CandidateResponse] | None = None


# ---------------------------------------------------------------------------
# 整定任务记录
# ---------------------------------------------------------------------------


class TuningTaskItem(CamelModel):
    """整定任务列表项。"""

    id: str
    loopId: str
    tagName: str | None = None
    modelType: ModelType
    modelParams: dict[str, Any] | None = None
    algorithm: TuningAlgorithm
    recommendedPid: dict[str, Any] | None = None
    fittingScore: float | None = None
    status: TuningTaskStatus
    createdBy: str | None = None
    createdAt: str
    # Phase 2.2 新增字段
    identifyMethod: IdentifyMethod | None = None
    dataSource: DataSource | None = None
    confidenceLevel: ConfidenceLevel | None = None
    confidenceReason: str | None = None
    excitationScore: float | None = None
    residualTestPassed: bool | None = None
    taskId: str | None = None
    completedAt: str | None = None


class TuningTaskDetail(TuningTaskItem):
    """整定任务详情。"""

    simulationResult: dict[str, Any] | None = None
    currentPid: dict[str, Any] | None = None
    # Phase 2.2 新增
    pidCandidates: dict[str, Any] | None = None
    candidateResults: dict[str, Any] | None = None


class CreateTuningTaskRequest(CamelModel):
    """创建整定任务（保存整定结果）。"""

    loopId: str
    modelType: ModelType
    modelParams: ModelParams
    algorithm: TuningAlgorithm
    recommendedPid: PidParams
    currentPid: PidParams | None = None
    fittingScore: float | None = None
    simulationResult: dict[str, Any] | None = None
    status: TuningTaskStatus = Field("SIMULATED", description="任务状态")
    # Phase 2.2 新增
    identifyMethod: IdentifyMethod | None = None
    dataSource: DataSource | None = None
    confidenceLevel: ConfidenceLevel | None = None
    confidenceReason: str | None = None
    excitationScore: float | None = None
    residualTestPassed: bool | None = None
    pidCandidates: dict[str, Any] | None = None
    candidateResults: dict[str, Any] | None = None


# ---------------------------------------------------------------------------
# 异步任务进度（Phase 2.2 新增）
# ---------------------------------------------------------------------------


class TaskProgress(CamelModel):
    """异步任务进度。"""

    taskId: str
    status: str = Field(..., description="任务状态: PENDING/RUNNING/SUCCESS/FAILED")
    progress: float = Field(0.0, description="进度 0~100")
    stage: str | None = Field(None, description="当前阶段: excitation/nonparametric/identify/...")
    message: str | None = None
    result: dict[str, Any] | None = None
    error: str | None = None


# ---------------------------------------------------------------------------
# 辨识片段预览（Phase 2.2 新增）
# ---------------------------------------------------------------------------


class IdentifySegment(CamelModel):
    """可辨识片段预览。"""

    startIdx: int
    endIdx: int
    mode: str | None = None
    excitationScore: float | None = None
    conditionNumber: float | None = None
    isSufficient: bool = False


class IdentifySegmentsRequest(CamelModel):
    """POST /tuning/identify/segments 请求体."""

    loopId: str
    startTime: str
    endTime: str


class IdentifySegmentsResult(CamelModel):
    """可辨识片段预览结果。"""

    loopId: str
    totalSegments: int
    segments: list[IdentifySegment]
    sufficientCount: int = Field(0, description="激励充分片段数")


# ---------------------------------------------------------------------------
# 整定效果统计
# ---------------------------------------------------------------------------


class TuningHistoryStats(CamelModel):
    """整定历史统计。"""

    totalTasks: int
    byAlgorithm: dict[str, int]
    byStatus: dict[str, int]
    avgFittingScore: float | None = None
    recentTasks: list[TuningTaskItem]


# ---------------------------------------------------------------------------
# 整定方法信息
# ---------------------------------------------------------------------------


class TuningMethodInfo(CamelModel):
    """整定方法信息。"""

    code: str
    name: str
    description: str
    applicableModel: str
    params: list[dict[str, Any]]
